### ISSUE TITLE
dnsdist corrupts terminal when run in foreground

### ISSUE THREAD

**Issue by philmayers** [PowerDNS/pdns#4733]
During a quick test, if I run dnsdist in the foreground, garbage is emitted to the terminal, usually rendering it unusable. For example:

```
# dnsdist --local 0.0.0.0:53 127.0.0.1:5300 -a 0.0.0.0/0
Unable to read configuration from '/etc/dnsdist/dnsdist.conf'
Listening on 0.0.0.0:53
dnsdist 1.1.0-beta1 comes with ABSOLUTELY NO WARRANTY. This is free software, and you are welcome to redistribute it according to the terms of the GPL version 2
ACL allowing queries from: 0.0.0.0/0
Added downstream server 127.0.0.1:5300
Marking downstream 127.0.0.1:5300 as 'up'
> &<garbled char>pixel
```

Everything after the ">" line above just appeared after a few seconds, I didn't type it. Using "script" to capture the terminal output to a file, "cat -A" says this was:

```
Added downstream server 127.0.0.1:5300^M$
Marking downstream 127.0.0.1:5300 as 'up'^M$
> ^@&M-_^@^A^@^@^A^@^@^@^@^@^@^Epixel^M$
quantserve^Ccom^@^@^A^@^APuTTY^M$
```

A quick strace -f -o of the process suggests it's a worker thread writing to fd 0 for some reason e.g.

```
30492 write(0, "\0+", 2)                = 2
30492 write(0, "G\312\1\0\0\1\0\0\0\0\0\0\n25-courier\4push\5app"..., 43) = 43
30492 read(0,  <unfinished ...>
```

30492 being a thread created earlier in the process by clone()

OS is CentOS 6 x64, package from the dnsdist repo page. Let me know if more info is required or if I'm doing something wrong.

**Comment by ahupowerdns** [comment]
Thanks for this report - which exact dnsdist repo do you use, 1.0 or git
master?

**Comment by philmayers** [comment]
Repo was the "dnsdist - version 1.1.X" one from the download page i.e.

```
https://repo.powerdns.com/repo-files/centos-dnsdist-11.repo
```

Just for clarity, the strace output from above is from an earlier run with the same symptoms hence the different domains in the stdout; also the "PuTTY" is obviously a terminal-specific thing.

Can retest with other versions if it's helpful

**Comment by philmayers** [comment]
Also RPM ended up being

Name        : dnsdist            
Version     : 1.1.0              
Release     : 0.1.beta1.1pdns.el6

**Comment by ahupowerdns** [comment]
Phil, 

Retesting with the "dnsdist - master branch" repo is most useful for us.
Could you try that?

Thanks!

**Comment by rgacogne** [comment]
The `strace` output suggests that the process writing to FD 0 is a TCP worker. I have no idea how it ends up writing on FD 0, so I would be very interested in seeing more of the `strace` outputs, ideally from the moment 30492 is created to the writes on FD 0.

**Comment by philmayers** [comment]
Sadly still seems to occur with dnsdist 0.0.1312g7541e94 from master repo

```
Marking downstream 127.0.0.1:5300 as 'up'^M$
> ^@+A<^A^@^@^A^@^@^@^@^@^@^M$
40-courier^Dpush^Eapple^Ccom^@^@^A^@^AP^M$
```

strace shows the same thing, seems to be one of the worker threads writing what looks like a raw DNS packet or possibly protobuf to fd 0

**Comment by rgacogne** [comment]
> strace shows the same thing, seems to be one of the worker threads writing what looks like a raw DNS packet or possibly protobuf to fd 0

Given the two-byte `write()`, I'm pretty sure it's a raw DNS packet over TCP.

**Comment by ahupowerdns** [comment]
a strace from startup with -ff would be super helpful. This would allow us
to observe the creation of the TCP sockets. Or not. Thanks for helping us
debug Phil!

**Comment by philmayers** [comment]
I'm a tad cautious about posting the full trace here due to privacy concerns related to the queries going through the box at the time (I can only repro this currently by making it briefly live!).

A slightly larger backtrace of the whole worker thread (main process elided to avoid privacy issues I'm afraid)

```
12531 set_robust_list(0x7ffaf4dfa9e0, 24 <unfinished ...>
12531 <... set_robust_list resumed> )   = 0
12531 read(11,  <unfinished ...>
12531 <... read resumed> "P\v\0\350\372\177\0\0", 8) = 8
12531 fcntl(17, F_GETFL)                = 0x2 (flags O_RDWR)
12531 fcntl(17, F_SETFL, O_RDWR|O_NONBLOCK) = 0
12531 getsockname(17, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("155.198.142.77")}, [16]) = 0
12531 read(17, "\0+", 2)                = 2
12531 read(17, "+\357\1\0\0\1\0\0\0\0\0\0\n12-courier\4push\5app"..., 43) = 43
12531 mmap(NULL, 134217728, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0) = 0x7ffad4000000
12531 munmap(0x7ffad8000000, 67108864)  = 0
12531 mprotect(0x7ffad4000000, 135168, PROT_READ|PROT_WRITE) = 0
12531 clock_gettime(CLOCK_MONOTONIC_RAW, {19192858, 278964650}) = 0
12531 clock_gettime(CLOCK_REALTIME, {1480865182, 417849641}) = 0
12531 socket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 19
12531 connect(19, {sa_family=AF_INET, sin_port=htons(5300), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 ECONNREFUSED (Connection refused)
12531 futex(0x7ffaffc6daf0, FUTEX_WAKE_PRIVATE, 2147483647) = 0
12531 close(17)                         = 0
12531 read(11,  <unfinished ...>
12531 <... read resumed> "\200\v\0\350\372\177\0\0", 8) = 8
12531 fcntl(17, F_GETFL)                = 0x2 (flags O_RDWR)
12531 fcntl(17, F_SETFL, O_RDWR|O_NONBLOCK) = 0
12531 getsockname(17, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("155.198.142.77")}, [16]) = 0
12531 read(17, "\0+", 2)                = 2
12531 read(17, "A<\1\0\0\1\0\0\0\0\0\0\n40-courier\4push\5app"..., 43) = 43
12531 clock_gettime(CLOCK_MONOTONIC_RAW, {19192890, 358844016}) = 0
12531 clock_gettime(CLOCK_REALTIME, {1480865214, 497744073}) = 0
12531 write(0, "\0+", 2)                = 2
12531 write(0, "A<\1\0\0\1\0\0\0\0\0\0\n40-courier\4push\5app"..., 43) = 43
12531 read(0, "uT", 2)                  = 2
12531 read(0,  <unfinished ...>
12531 <... read resumed> "TY", 30036)   = 2
12531 read(0,  <unfinished ...>
12531 +++ killed by SIGINT +++
```

Interesting that the issue seems to happen after an ECONNREFUSED to the backend server 155.198.142.77 - not no socket() call in the 2nd attempt/chunk.

I've got no idea where it's getting fd 0 from; the whole strace confirms that fd 0 isn't being accidentally closed so socket() should never return it, which means SSocket and thus setupTCPDownstream should not return 0 and it shouldn't be in map<> sockets and thus dsock. Very odd...

GDB backtrace of that worker thread from another instance in case it's useful

```
#0  write () at ../sysdeps/unix/syscall-template.S:82
#1  0x00007ffff7ebf5ef in writen2WithTimeout (fd=0, buffer=0x7ffff213b72e, len=2, timeout=30) at misc.cc:146
#2  0x00007ffff7e9720e in putNonBlockingMsgLen (fd=Unhandled dwarf expression opcode 0xf3
) at dnsdist-tcp.cc:117
#3  0x00007ffff7e996ba in sendNonBlockingMsgLen (pipefd=15) at dnsdist-tcp.cc:128
#4  tcpClientThread (pipefd=15) at dnsdist-tcp.cc:367
#5  0x00007ffff6bbb470 in std::(anonymous namespace)::execute_native_thread_routine (__p=<value optimized out>) at ../../../../libstdc++-v3/src/thread.cc:44
#6  0x00007ffff6455aa1 in start_thread (arg=0x7ffff213d700) at pthread_create.c:301
#7  0x00007ffff61a293d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:115
```

The version of GDB I have can't decode all the DWARF symbols in the backtrace or deref any of the std:: values from the core file so it's a bit useless beyond this I'm afraid.

Unfortunately I've reached the limits of the disruption I can sneak in this evening and will have to stop now - I'll try and repro on the bench tomorrow.

**Comment by ahupowerdns** [comment]
That is some super debugging Phil! Braining on it now. But thank you so much.

**Comment by ahupowerdns** [comment]
Ok, so if SConnect throws, we may have already created an empty sockets[dsremote] which will indeed have a zero-default initialized value in there for the socket. So the code should be two-step, first get the fd and only THEN assign it. @rgacogne does this agree with your understanding?

**Comment by rgacogne** [comment]
Yes, it makes total sense! Proposed fix in #4734.
